crates/runix/src/uri_parser.rs

```diff
@@ -0,0 +1,971 @@
+    pub attrs: Attrs,
+    pub string: String,
+}
+
+/// A strongly-typed flake reference that has already gone through deserialization and validation
+#[derive(Debug, Clone)]
+pub struct ParsedFlakeReference {
+    pub attrs: Attrs,
+    pub string: String,
+    pub flake_type: FlakeType,
+}
+
+impl TryFrom<DeserializedFlakeRef> for ParsedFlakeReference {
+    type Error = UriParseError;
+
+    fn try_from(flake: DeserializedFlakeRef) -> Result<Self, Self::Error> {
+        use FlakeType::*;
+        let Some(raw_flake_type) = flake.attrs.get("type") else {
+            return Err(UriParseError::TypeNotFound);
+        };
+        let Value::String(raw_flake_type) = raw_flake_type else {
+            return Err(UriParseError::TypeWasWrongType(raw_flake_type.clone()));
+        };
+        let protocol = url_scheme(flake.attrs.get("url"));
+        match raw_flake_type.as_ref() {
+            "git" | "mercurial" | "tarball" | "file" => {
+                if protocol.is_none() {
+                    return Err(UriParseError::ProtocolNotFound);
+                }
+            },
+            _ => {},
+        };
+        // Unwrapping `protocol` is now safe because we've verified that it's `Some` for any flake type that needs it
+        let flake_type = match raw_flake_type.as_ref() {
+            "path" => Path,
+            "github" => Github,
+            "gitlab" => Gitlab,
+            "sourcehut" => Sourcehut,
+            "indirect" => Indirect,
+            "git" => {
+                let p_type = GitProtocolType::from_str(protocol.unwrap().as_ref())?;
+                Git(p_type)
+            },
+            "mercurial" => {
+                let p_type = MercurialProtocolType::from_str(protocol.unwrap().as_ref())?;
+                Mercurial(p_type)
+            },
+            "tarball" => {
+                let p_type = TarballProtocolType::from_str(protocol.unwrap().as_ref())?;
+                Tarball(p_type)
+            },
+            "file" => {
+                let p_type = FileProtocolType::from_str(protocol.unwrap().as_ref())?;
+                File(p_type)
+            },
+            _ => return Err(UriParseError::UnrecognizedFlakeType(raw_flake_type.clone())),
+        };
+        Ok(ParsedFlakeReference {
+            attrs: flake.attrs,
+            string: flake.string,
+            flake_type,
+        })
+    }
+}
+
+/// Extracts the scheme out of a flake reference URL
+fn url_scheme(url: Option<&Value>) -> Option<String> {
+    let Some(Value::String(url)) = url else {
+        return None;
+    };
+    url.split(":/").next().map(|s| s.to_string())
+}
+
+/// The various types of flake references
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum FlakeType {
+    Path,
+    Git(GitProtocolType),
+    Mercurial(MercurialProtocolType),
+    Tarball(TarballProtocolType),
+    File(FileProtocolType),
+    Github,
+    Gitlab,
+    Sourcehut,
+    Indirect,
+}
+
+impl From<FlakeType> for String {
```

Comment:
Why not `Display`, which can also be derived from the type directly

Reply:
This is actually just so that I can do the type conversion `FlakeType -> String` for comparison in the `parses_test_bank` test. If this was meant for user-facing display I would use `Display` and format the service names properly e.g. `"github" -> "GitHub"`.